Alter Table Add Constraint

velioglu opened the pull request:
Fixes #1239.

Follow the same logic for create table and alter table to propagate constraint definitions to workers.

velioglu commented:
There is a problem on the implementation when the constraints are checked with ErrorIfNotSupportedConstraint from multi_ProcessUtility. Transaction errors out from the last check of ErrorIfNotSupportedForeignConstraint with the error message of :

ERROR:  cannot create foreign key constraint
DETAIL:  Citus Community Edition currently supports foreign key constraints only for "citus.shard_replication_factor = 1".
HINT:  Please change "citus.shard_replication_factor to 1". To learn more about using foreign keys with other replication factors, please contact us at https://citusdata.com/about/contact_us.

When I check the shard replication factor from the error it is equal to 2, yet when I control it using SHOW citus.shard_replication_factor via sql it is 1. Note that, this errors occurs with create_distributed_table also.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,131 @@
+ALTER TABLE unique_test_table DROP CONSTRAINT unn_id;
+-- Can create unique constraint over multiple columns including
+-- distribution column
+INSERT INTO unique_test_table VALUES(1, 'Mehmet');
```
Let's insert this just before the other insertion

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,131 @@
+--
```
Let's also add test cases when we fail to create constraints because of;
* missing constraint name
* ADD CONSTRAINT is not only subcommand
* Constraint not on distribution column
* other...

Also let's add test for reference/append distributed tables too

byucesoy replied:
We should also add tests with transactions

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,131 @@
+DETAIL:  Key (product_no)=(1) already exists.
+CONTEXT:  while executing command on localhost:57638
+drop table products;
+-- Check "EXCLUSION CONSTRAINT"
```
And also tests for NOT NULL

velioglu commented on src/backend/distributed/commands/multi_copy.c:
```diff
@@ -810,10 +810,11 @@ OpenCopyConnections(CopyStmt *copyStatement, ShardConnections *shardConnections,
- * If it finds such column, that means we cannot use binary format for COPY, because
- * binary format sends Oid of the types, which are generally not same in master and
- * worker nodes for user-defined types.
+ * and looks for a column whose type is array of user-defined type. If it finds such
```
oops 😃

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,459 @@
+ERROR:  single-shard DML commands must not appear in transaction blocks which contain multi-shard data modifications
+DROP TABLE products;
+ERROR:  current transaction is aborted, commands ignored until end of transaction block
+END;
```
Also let's have a test

```sql
BEGIN:
-- create a distributed table
-- alter table add constraint
ROLLBACK;
-- see no constraint is created in master
-- see no constraint is created in one of the worker
```

marcocitus commented on src/test/regress/sql/multi_alter_table_add_constraints.sql:
```diff
@@ -0,0 +1,387 @@
+SELECT create_distributed_table('unique_test_table_append', 'id', 'append');
+
+-- We can add unique constraint on any column with reference tables
+ALTER TABLE unique_test_table_append ADD CONSTRAINT unn_name UNIQUE(name);
```
We should prevent it. Append-distributed tables may have overlapping shards and typically do.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,459 @@
+
+-- We can add unique constraint on any column with reference tables
+ALTER TABLE unique_test_table_append ADD CONSTRAINT unn_name UNIQUE(name);
+WARNING:  table "unique_test_table_append" has a UNIQUE or EXCLUDE constraint
```
We discussed about this with @marcocitus. We want to prevent this behavior but also we do not want to break existing environments. If user has a append distributed table with unique constraint, we should not error out when creating new shard for this table (with COPY for example). 

It seems there are four case to consider. Correct behavior for each case should be;
- Table is not distributed yet, user creates an unique constraint, then distributes by append -> **error out during distributed table creation**
- Table is distributed(before this PR) by append and has NO unique constraint, user creates unique constraint with ALTER TABLE -> **error out during ALTER TABLE**
- Table is distributed(before this PR) by append and has unique constraint, user replicates a shard -> **propagate unique constraints to new shards**
- Table is distributed(before this PR) by append and has unique constraint, user creates new shard -> **propagate unique constraints to new shards**

@metdos we can implement this behavior with one line fix, do you have any objection adding the fix to this PR.

metdos commented:
> We discussed about this with @marcocitus. We want to prevent this behavior but also we do not want to break existing environments. If user has a append distributed table with unique constraint, we should not error out when creating new shard for this table (with COPY for example).
> 
> It seems there are four case to consider. Correct behavior for each case should be;
> 
> Table is not distributed yet, user creates an unique constraint, then distributes by append -> error out during distributed table creation
> Table is distributed(before this PR) by append and has NO unique constraint, user creates unique constraint with ALTER TABLE -> error out during ALTER TABLE
> Table is distributed(before this PR) by append and has unique constraint, user replicates a shard -> propagate unique constraints to new shards
> Table is distributed(before this PR) by append and has unique constraint, user creates new shard -> propagate unique constraints to new shards
> @metdos we can implement this behavior with one line fix, do you have any objection adding the fix to this PR.

We can add this as a separate commit in this PR and also add your explanation to commit description. Then, we should be okay!

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+
+-- Can only add unique constraint on distribution column (or group
+-- of columns including distribution column)
+ALTER TABLE unique_test_table ADD CONSTRAINT unn_name UNIQUE(name);
```
Specify that the command will error out.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ALTER TABLE unique_test_table DROP CONSTRAINT unn_id;
+-- Can create unique constraint over multiple columns which must include
+-- distribution column
+INSERT INTO unique_test_table VALUES(1, 'Mehmet');
```
According to comment, I expect that next example will be successfully adding a unique constraints over multiple columns, but we test failing case when there is already conflicting rows in the table. Let's keep both tests, but write separate comments for them.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ERROR:  cannot add UNIQUE, PRIMARY KEY or EXCLUDE constraint on "products_append" 
+DETAIL:  UNIQUE constraints, EXCLUDE constraints, and PRIMARY KEYs on append-partitioned tables cannot be enforced.
+HINT:  Consider using hash partitioning.
+COPY products_append FROM STDIN DELIMITER AS ',';
```
No need to COPY data.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ 
+(1 row)
+
+-- Can add column and table constraints
```
This comment does not make sense.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+-- Can add column and table constraints
+ALTER TABLE products ADD CONSTRAINT p_check CHECK(price > 0);
+ALTER TABLE products ADD CONSTRAINT p_multi_check CHECK(price > discounted_price);
+-- Error out from the related shard. Table can not have row conflicting with p_multi_check
```
"Related shard" term does not feel right here. Lets make this (and others) something similar to;

"This query will error out because insterting row's price is less than 0.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ 
+(1 row)
+
+-- Can add column and table constraints
```
Similar to above

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ALTER TABLE products ADD CONSTRAINT exc_name EXCLUDE USING btree (name with =); 
+ERROR:  cannot create constraint on "products"
+DETAIL:  Distributed relations cannot have UNIQUE, EXCLUDE, or PRIMARY KEY constraints that do not include the partition column (with an equality operator if EXCLUDE).
+-- We can add composite exclusion because none of pair of rows are conflicting
```
We can add composite exclusion -> corrent
because none of pair of rows are conflicting -> did not understand that

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+
+-- Can only add exclusion constraint on distribution column (or group of columns
+-- including distribution column)
+-- Command below should error out since 'name' is not a distribution column
```
Good example of comment :+1:

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ALTER TABLE products_ref ADD CONSTRAINT exc_name EXCLUDE USING btree (name with =); 
+-- We can add composite exclusion because none of pair of rows are conflicting
+ALTER TABLE products_ref ADD CONSTRAINT exc_pno_name EXCLUDE USING btree (product_no with =, name with =);
+-- Third insertion will error out, since it has the same name with second insertion
```
Good example of comment :+1:

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+(1 row)
+
+-- Cannot add exclusion constraint on append table
+-- Command below should error out since 'name' is not a distribution column
```
Erroring out reason is not correct.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+-- Should error out since conflicts with the above single-shard data modification command.
+ALTER TABLE products ADD CONSTRAINT unn_pno UNIQUE(product_no);
+ERROR:  distributed DDL commands must not appear within transaction blocks containing single-shard data modifications
+COMMIT;
```
After error we ROLLBACK, not COMMIT.

byucesoy commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,501 @@
+ALTER TABLE products ADD CONSTRAINT check_price CHECK(price > discounted_price);
+ALTER TABLE products ADD CONSTRAINT p_key_product PRIMARY KEY(product_no);
+ROLLBACK;
+-- All of them should be equal to zero
```
Instead of querying pg_constraint, you can just use

/d products and see there are no constraints.

byucesoy commented on src/test/regress/expected/multi_complex_expressions.out:
```diff
@@ -445,15 +445,15 @@ DEBUG:  push down of limit count: 30
 -----------+-----------+------------
     154380 |       421 |      26.00
-    103981 |       325 |      26.00
+    138854 |        79 |      26.00
```
Why these numbers changed? I did not understand their relation with constraints on append table.

velioglu replied:
Since removing PRIMARY KEY constraints change the query plan, it effects the order of rows with the same l_quantity. I have changed the column used by ORDER BY to l_partkey, o_custkey from l_quantity to generate the same result regardless of the plan.

byucesoy commented on src/test/regress/expected/multi_create_table_constraints.out:
```diff
@@ -6,13 +6,10 @@ ALTER SEQUENCE pg_catalog.pg_dist_jobid_seq RESTART 365000;
 CREATE TABLE uniq_cns_append_tables
 (
-	partition_col integer UNIQUE,
+	partition_col integer,
```
According to comment above, this(and below) table is used to test that Citus forbids unique and EXCLUDE constraints on append-partitioned tables.

So I think it is better to keep UNIQUE keyword and see error message. It seems that this table is not used anywhere else.

byucesoy commented on src/test/regress/input/multi_copy.source:
```diff
@@ -341,8 +341,7 @@ CREATE TABLE customer_worker_copy_append (
         c_mktsegment char(10),
-        c_comment varchar(117),
-                primary key (c_custkey));
+        c_comment varchar(117));
```
Fix the comment above.

byucesoy commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -133,6 +136,11 @@ static void ErrorIfDistributedAlterSeqOwnedBy(AlterSeqStmt *alterSeqStmt);
 static bool OptionsSpecifyOwnedBy(List *optionList, Oid *ownedByTableId);
 static void ErrorIfDistributedRenameStmt(RenameStmt *renameStatement);
+static void ErrorIfUnsupportedAlterAddConstraintStmt(AlterTableStmt *alterTableStatement);
+static void ErrorIfNotSupportedForeignConstraint(Relation relation,
```
I know this is existing function, but I realized we generally used "ErrorIf**Un**supported..." instead of "ErrorIf**Not**Supporter...", could you rename this and ErrorIfNotSupportedConstraint functions.

byucesoy commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -391,6 +399,27 @@ multi_ProcessUtility(Node *parsetree,
+		 */
+		if (IsA(parsetree, AlterTableStmt))
+		{
+			AlterTableStmt *alterTableStatement = (AlterTableStmt *) parsetree;
```
Let's add here an Assert to ensure commandList really has single sub-command, similar to [here](https://github.com/citusdata/citus/blob/master/src/backend/distributed/executor/multi_utility.c#L944)

metdos commented on src/test/regress/expected/multi_complex_expressions.out:
```diff
@@ -439,21 +439,21 @@ FROM
 	li.l_quantity > 25
 ORDER BY
-	li.l_quantity
+	li.l_partkey, o.o_custkey
```
You can do
```sql
ORDER BY
    li.l_quantity, li.l_partkey, o.o_custkey
```

metdos commented on src/test/regress/expected/multi_create_table_constraints.out:
```diff
@@ -41,15 +32,15 @@ CREATE TABLE pk_on_non_part_col
 );
 SELECT master_create_distributed_table('pk_on_non_part_col', 'partition_col', 'hash');
-ERROR:  cannot distribute relation: "pk_on_non_part_col"
+ERROR:  cannot create constraint on "pk_on_non_part_col"
```
Why did we change this error message? Th older one makes more sense.

velioglu replied:
We check the constraints of CREATE TABLE and ALTER TABLE statements using the same part of the code. For the ALTER TABLE case, "cannot distribute relation" seems meaningless that the table has already distributed.

metdos commented on src/test/regress/expected/multi_foreign_key.out:
```diff
@@ -464,8 +462,7 @@ SELECT create_distributed_table('referencing_table', 'ref_id', 'hash');
 ALTER TABLE referencing_table ADD CONSTRAINT test_constraint FOREIGN KEY(ref_id) REFERENCES referenced_table ON DELETE CASCADE;
-ERROR:  cannot create foreign key constraint because referenced column list is empty
-HINT:  Add column names to "REFERENCES" part of the statement.
+ERROR:  number of referencing and referenced columns for foreign key disagree
```
The error message seems more explanatory.

velioglu replied:
Error message is generated by Postgres, i.e. user will be faced with the same error message running the same command without distributing tables.

`CREATE TABLE referenced_table(id int UNIQUE, test_column int, PRIMARY KEY(id, test_column));
CREATE TABLE referencing_table(id int, ref_id int);
ALTER TABLE referencing_table ADD CONSTRAINT test_constraint FOREIGN KEY(ref_id) REFERENCES referenced_table ON DELETE CASCADE;
ERROR:  number of referencing and referenced columns for foreign key disagree`

metdos commented on src/test/regress/expected/multi_foreign_key.out:
```diff
@@ -882,8 +870,7 @@ SELECT create_reference_table('reference_table');
 ALTER TABLE reference_table ADD CONSTRAINT fk FOREIGN KEY(referencing_column) REFERENCES referenced_local_table(id);
-ERROR:  cannot create foreign key constraint
-DETAIL:  Foreign key constraints are not allowed from or to reference tables.
+ERROR:  relation referenced_local_table is not distributed
```
This new error message also is not a bit confusing.

velioglu replied:
Since we check foreign table constraints after we create them, planning phase catch the not distributed table before constraint checking catch the reference table. I can get the change back by adding a check of reference tables before planning phase, yet it will break the logic of checking constraints of ALTER TABLE and CREATE TABLE statements using the same code part.

metdos commented:
@velioglu, I have a question. I see that we moved `ErrorIfNotSupportedConstraint` and `ErrorIfNotSupportedConstraint` to another file. Did we just move them or did we also make some changes on them?

velioglu commented:
@metdos, we almost just moved and renamed them. Only changed some error messages to make them meaningful with both CREATE TABLE and ALTER TABLE statements and removed an unnecessary check for co-located tables in ErrorIfUnsupportedForeignConstraint.

metdos commented:
> @metdos, we almost just moved and renamed them. Only changed some error messages to make them meaningful with both CREATE TABLE and ALTER TABLE statements and removed an unnecessary check for co-located tables in ErrorIfUnsupportedForeignConstraint.

What about here?
https://github.com/citusdata/citus/pull/1292/files#diff-a8033caa333591e68f4fb7a599757296R2048

metdos commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1623,211 +1651,421 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+									  "be in the same ordinal in both sides.")));
+		}
+
+		multiShardedReferencingExist = IsDistributedTable(referencingTableId) &&
```
This part is confusing.  I assume `multiShardedReferencingExist` means `IsDistributedTable(referencingTableId)` and `multiShardedReferencingBeingCreated ` means `!IsDistributedTable(referencingTableId)`. `!SingleReplicatedTable(referencingTableId)` and `ShardReplicationFactor > 1` are only exist to use in if below.

What are you trying to do here? I presume you want to distinguish between `CREATE TABLE` and `ALTER TABLE` and check `shard_count > 1` according to that. If we rewrite this in this form, will it be correct?

```c
if (IsDistributedTable(referencingTableId))
  {
    \\ ALTER TABLE...
    if (!SingleReplicatedTable(referencingTableId))
    {
        singleReplicatedTable = false;
     }
}
else
{
  \\ CREATE TABLE
  if (ShardReplicationFactor > 1)
  {
       singleReplicatedTable = false;
  }
}

if(!singleReplicatedTable)
{
 \\ error out
}
```

You may also want to use `selfReferencingTable` in checks above. What is our policy for self-referencing tables? Do we support constraints on them all the time?

Does this make sense? What do you think?

velioglu replied:
Yes, I was trying to distinguish whether the function is called for CREATE TABLE or ALTER TABLE by IsDistributedTable(), so if we rewrite it in the form you have written that will also be correct. Self-referencing tables also error out if the shard replication factor is > 1. 

What we check by multiShardedReferencingExist and multiShardedReferencingBeingCreated are not clear enough now, so I'm going to change the code to make it more readable.

metdos commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1623,211 +1649,432 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+		}
+
+		/* check whether creating single replicated table with foreign constraint */
+		if (!IsDistributedTable(referencingTableId))
```
We can use `else` instead of this `if`. You can get comments inside if blocks to have symmetry.

metdos commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,477 @@
+-- Can create unique constraint over multiple columns which must include
+-- distribution column
+ALTER TABLE unique_test_table ADD CONSTRAINT unn_id_name UNIQUE(id, name);
+-- Error out, since tables can not have two rows woth same id and name.
```
typo: woth

metdos commented on src/test/regress/expected/multi_alter_table_add_constraints.out:
```diff
@@ -0,0 +1,477 @@
+(1 row)
+
+ALTER TABLE products ALTER COLUMN name SET NOT NULL;
+-- Insertions will error our since both product_no and name can not have NULL value
```
typo: our

metdos commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1583,7 +1610,6 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
 			case AT_DropColumn:
 			case AT_ColumnDefault:
 			case AT_AlterColumnType:
-			case AT_SetNotNull:
```
Why did move `AT_SetNotNull` to below?

velioglu replied:
I think other 4 constraints should be prevented to apply them over the partition column, but adding NOT NULL constraint over the partition column won't cause a problem.

metdos commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -391,6 +399,25 @@ multi_ProcessUtility(Node *parsetree,
 	{
 		ListCell *ddlJobCell = NULL;
 
+		if (IsA(parsetree, AlterTableStmt))
```
Should not we have this check in `ErrorIfUnsupportedAlterTableStmt`?

Looking this more, it looks like you do checks here after running master query so it becomes easier to use create table checks? Am I right?

velioglu replied:
Yes, you are right.

metdos replied:
Can we add a comment explaining why we have these checks here instead of earlier with other checks?

velioglu commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -391,6 +399,25 @@ multi_ProcessUtility(Node *parsetree,
+
+				if (alterTableType == AT_AddConstraint)
+				{
+					Assert(commandList->length <= 1);
```
Since we checked that length of commandList is smaller than 1 on ErrorIfUnsupportedAlterTableStmt, we are sure list_length(commandList) = 1. Both ErrorIfUnsupportedAlterTableStmt and PlanAlterTableStmt check the length of command list in that way, should I fix them too ?

metdos replied:
> Since we checked that length of commandList is smaller than 1 on ErrorIfUnsupportedAlterTableStmt, we are sure list_length(commandList) = 1.

Nice!

> Both ErrorIfUnsupportedAlterTableStmt and PlanAlterTableStmt check the length of command list in that way, should I fix them too ?

I think we can change `->length` to `list_length` because `list_length` makes null check. Even though it is not required here, it's a better practice.

metdos commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1532,7 +1559,7 @@ ErrorIfUnsupportedDropIndexStmt(DropStmt *dropIndexStatement)
  * ALTER TABLE SET|DROP NOT NULL
  * ALTER TABLE SET|DROP DEFAULT
- * ALTER TABLE ADD|DROP CONSTRAINT FOREIGN
+ * ALTER TABLE ADD|DROP CONSTRAINT
```
This my mistake.  CONSTRAINT >  CONSTRAINT FOREIGN

metdos commented:
Looks good 🚀
